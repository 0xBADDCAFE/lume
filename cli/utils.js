import { existsSync } from "../deps/fs.js";
import lume from "../mod.js";
import { join, resolve } from "../deps/path.js";
import { brightGreen } from "../deps/colors.js";

/**
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options, site) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  if (!site) {
    if (existsSync(configFile)) {
      const mod = await import(`file://${configFile}`);
      site = mod.default;
    } else {
      site = lume();
    }
  }

  site.options.cwd = options.root;

  if (options.dev) {
    site.options.dev = options.dev;
  }

  if (options.location) {
    site.options.location = new URL(options.location);
  }

  if (options.src) {
    site.options.src = options.src;
  }

  if (options.dest) {
    site.options.dest = options.dest;
  }

  if (options["--"]) {
    site.options.flags = options["--"];
  }

  return site;
}

/**
 * @param command name of the command you're validating
 * @param args array of cli arguments (i.e. options._)
 * @param max the max number of args
 * @param min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}
    
    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}
