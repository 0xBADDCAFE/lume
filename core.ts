import type {
  ComponentsOptions,
  default as Site,
  ServerOptions,
  SiteEvent,
  SiteEventType,
  SiteOptions,
  WatcherOptions,
} from "./core/site.ts";

import type {
  default as Events,
  Event,
  EventListener,
  EventOptions,
} from "./core/events.ts";

import type {
  default as Scripts,
  ScriptOptions,
  ScriptOrFunction,
} from "./core/scripts.ts";

import type {
  default as Reader,
  DirEntry,
  FileInfo,
  Loader,
} from "./core/reader.ts";
import type Logger from "./core/logger.ts";
import type Writer from "./core/writer.ts";
import type IncludesLoader from "./core/includes_loader.ts";
import type { default as PageLoader } from "./core/page_loader.ts";
import type DataLoader from "./core/data_loader.ts";
import type ComponentLoader from "./core/component_loader.ts";
import type {
  Component,
  Components,
  Content,
  Data,
  Dest,
  Directory,
  Page,
  Src,
  StaticFile,
} from "./core/filesystem.ts";
import type {
  ComponentFunction,
  default as Source,
  ProxyComponents,
} from "./core/source.ts";
import type {
  default as Renderer,
  Engine,
  Helper,
  HelperOptions,
} from "./core/renderer.ts";
import type { default as Processors, Processor } from "./core/processors.ts";
import type { default as Scopes, ScopeFilter } from "./core/scopes.ts";
import type { ErrorData, Exception } from "./core/errors.ts";
import type { default as Formats, Format } from "./core/formats.ts";
import type {
  default as Server,
  Middleware,
  RequestHandler,
  ServerEvent,
  ServerEventType,
} from "./core/server.ts";
import type {
  default as Watcher,
  WatchEvent,
  WatchEventType,
} from "./core/watcher.ts";

// Plugins types
import type { PaginationInfo, Paginator } from "./plugins/paginate.ts";
import type { Transformation, Transformations } from "./plugins/imagick.ts";
import type { MetaData } from "./plugins/metas.ts";
import type { Search } from "./plugins/search.ts";

/** The method that installs a plugin */
type PluginSetup = (options: unknown) => Plugin;

/** A generic Lume plugin */
type Plugin = (site: Site) => void;

export type {
  Component,
  ComponentFunction,
  ComponentLoader,
  Components,
  ComponentsOptions,
  Content,
  Data,
  DataLoader,
  Dest,
  Directory,
  DirEntry,
  Engine,
  ErrorData,
  Event,
  EventListener,
  EventOptions,
  Events,
  Exception,
  FileInfo,
  Format,
  Formats,
  Helper,
  HelperOptions,
  IncludesLoader,
  Loader,
  Logger,
  Middleware,
  Page,
  PageLoader,
  Plugin,
  PluginSetup,
  Processor,
  Processors,
  ProxyComponents,
  Reader,
  Renderer,
  RequestHandler,
  ScopeFilter,
  Scopes,
  ScriptOptions,
  ScriptOrFunction,
  Scripts,
  Server,
  ServerEvent,
  ServerEventType,
  ServerOptions,
  Site,
  SiteEvent,
  SiteEventType,
  SiteOptions,
  Source,
  Src,
  StaticFile,
  Watcher,
  WatcherOptions,
  WatchEvent,
  WatchEventType,
  Writer,
};

export interface PageData extends Data {
  /** The title of the page */
  title?: string;

  /** The page url */
  url: string;

  /** The available components */
  // deno-lint-ignore no-explicit-any
  comp?: any;

  /** Plugin Paginator: The paginator helper */
  paginate?: Paginator<Page>;

  /** Plugin Pagination: The pagination info */
  pagination?: PaginationInfo;

  /** Plugin Pagination: The pagination result */
  results?: Page[];

  /** Plugin Imagick: Image transformations */
  imagick?: Transformation | Transformations;

  /** Plugin Metas: Meta elements */
  metas?: MetaData;

  /** Plugin NetlifyCMS: CMS configuration */
  netlify_cms?: Record<string, unknown>;

  /** Plugin Search: The searcher helper */
  search?: Search;
}
