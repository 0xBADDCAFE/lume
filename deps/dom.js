export * from "https://raw.githubusercontent.com/b-fuze/deno-dom/master/deno-dom-wasm.ts";
