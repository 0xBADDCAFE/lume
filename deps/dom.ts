export * from "https://deno.land/x/deno_dom@v0.1.27-alpha/deno-dom-wasm.ts";
