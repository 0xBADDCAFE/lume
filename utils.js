import { bold, red } from "./deps/colors.js";
import { SEP } from "./deps/path.js";

export async function concurrent(iterable, iteratorFn, limit = 200) {
  const executing = [];

  for await (const item of iterable) {
    const p = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

export const mimes = new Map([
  [".html", "text/html; charset=utf-8"],
  [".txt", "text/plain; charset=utf-8"],
  [".js", "text/javascript; charset=utf-8"],
  [".css", "text/css; charset=utf-8"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".json", "application/json"],
  [".webmanifest", "application/manifest+json"],
  [".ico", "image/x-icon"],
  [".avif", "image/avif"],
  [".png", "image/png"],
  [".apng", "image/apng"],
  [".jpg", "image/jpg"],
  [".jpeg", "image/jpg"],
  [".gif", "image/gif"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mpeg"],
  [".xml", "text/xml"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".wasm", "application/wasm"],
  [".webp", "image/webp"],
  [".webm", "video/webm"],
  [".zip", "application/zip"],
]);

export function error(context, message, exception) {
  console.error(bold(red(`${context}:`)), message);
  if (exception) {
    console.error(exception);
  }
  console.error("");
}

export function merge(defaults, user) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (
      typeof merged[key] === "object" && typeof value === "object" &&
      !Array.isArray(merged[key])
    ) {
      merged[key] = merge(merged[key], value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

export function normalizePath(path) {
  // Is Windows path
  if (SEP !== "/") {
    return path.replaceAll(SEP, "/");
  }

  return path;
}

export const slugChars = new Map([
  ["รฐ", "d"],
  ["ร", "ss"],
]);

export function slugify(string) {
  return string.toLowerCase()
    .normalize("NFKD")
    .replaceAll(/[\s_-]+/g, "-")
    .replaceAll(/[^\w\/-]/g, (char) => slugChars.get(char) || "");
}

export function searchByExtension(path, extensions) {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}
