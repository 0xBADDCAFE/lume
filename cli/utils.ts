import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { dim, red } from "../deps/colors.ts";
import { getConfigFile } from "../core/utils.ts";

import type { Site } from "../core.ts";

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  const path = await getConfigFile(root, config);

  if (path) {
    console.log(`Loading config file ${dim(path)}`);
    console.log();
    const mod = await import(toFileUrl(path).href);
    return mod.default;
  }

  return lume();
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nano_jsx",
  "netlify_cms",
  "on_demand",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "resolve_urls",
  "sass",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}
