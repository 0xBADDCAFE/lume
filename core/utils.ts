import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import {
  brightGreen,
  brightYellow,
  cyan,
  dim,
  green,
  red,
} from "../deps/colors.ts";
import { dirname, extname, join, posix, SEP } from "../deps/path.ts";
import { parse } from "../deps/jsonc.ts";
import { Exception } from "./errors.ts";

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "jsx_preact",
  "katex",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "netlify_cms",
  "on_demand",
  "pagefind",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "remark",
  "resolve_urls",
  "sass",
  "sheets",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

/** A list of the available plugins with init configurations */
export const initPlugins = [
  "jsx",
  "jsx_preact",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}

/** Check the compatibility with the current Deno version */
export function checkDenoVersion(): void {
  const minimum = "1.25.4";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}

/** Check if the current version is outdated */
export async function checkUpgrade(): Promise<void> {
  const current = getLumeVersion();

  // It's a local version
  if (current.startsWith("local ")) {
    return;
  }

  const stable = !!current.match(/^v\d+\./);
  const expires = 1000 * 60 * 60 * 24; // 1 day
  const interval = localStorage.getItem("lume-upgrade");

  if (interval && parseInt(interval) + expires > Date.now()) {
    return;
  }

  localStorage.setItem("lume-upgrade", Date.now().toString());

  const latest = stable
    ? await getLatestVersion()
    : await getLatestDevelopmentVersion();

  if (current === latest) {
    return;
  }

  let global = false;
  try {
    await Promise.any([
      Deno.stat("deno.json"),
      Deno.stat("deno.jsonc"),
    ]);
    await Deno.stat("import_map.json");
  } catch {
    global = true;
  }

  const command = global
    ? (stable ? "lume upgrade --global" : "lume upgrade --dev --global")
    : (stable ? "deno task lume upgrade" : "deno task lume upgrade --dev");

  log(
    `Update available ${dim(current)}  â†’ ${green(latest)}`,
    `Run ${cyan(command)} to update`,
  );
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    "https://api.github.com/repos/lumeland/lume/commits?per_page=1",
  );
  const commits = await response.json();
  return commits[0].sha;
}

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx";
    jsxImportSource?: string;
  };
  [key: string]: unknown;
}

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Encode a message using SHA-1 algorithm */
export async function sha1(message: string | Uint8Array): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Helper to create optional properties recursively */
// deno-lint-ignore ban-types
export type DeepPartial<T> = T extends object ? {
    [P in keyof T]?: DeepPartial<T[P]>;
  }
  : T;

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type> | DeepPartial<Type>,
): Type {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (value === undefined) {
      continue;
    }

    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

const reactElement = Symbol.for("react.element");
const objectConstructor = {}.constructor;

/** Check if the argument passed is a plain object */
export function isPlainObject(obj: unknown): obj is Record<string, unknown> {
  return typeof obj === "object" && obj !== null &&
    obj.constructor === objectConstructor &&
    // @ts-ignore: Check if the argument passed is a React element
    obj["$$typeof"] !== reactElement;
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/")
 * and ensure it starts with "/".
 */
export function normalizePath(path: string) {
  if (SEP !== "/") {
    path = path.replaceAll(SEP, "/");

    // Is absolute Windows path (C:/...)
    if (path.includes(":/")) {
      return path;
    }
  }

  return posix.join("/", path);
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return `<!DOCTYPE html>\n${documentElement?.outerHTML || ""}`;
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/** Return the current installed version */
export function getLumeVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Returns the _config file of a site */
export async function getConfigFile(
  path?: string,
): Promise<string | undefined> {
  if (path) {
    try {
      return await Deno.realPath(path);
    } catch (cause) {
      throw new Exception("Config file not found", { path, cause });
    }
  }

  const paths = ["_config.js", "_config.ts"];

  for (const path of paths) {
    try {
      return await Deno.realPath(path);
    } catch {
      // Ignore
    }
  }
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(
  importMapFile?: string,
): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      importMapFile ||= config.importMap;
      if (importMapFile) {
        config.importMap = importMapFile;
        const importMap: ImportMap = isUrl(importMapFile)
          ? await (await fetch(importMapFile)).json()
          : await JSON.parse(await Deno.readTextFile(importMapFile));

        return { file, config, importMap };
      }
      return { file, config };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}

/** Update the Deno configuration */
export async function writeDenoConfig(options: DenoConfigResult) {
  const { file, config, importMap } = options;

  if (importMap && !config.importMap) {
    config.importMap = "./import_map.json";
  }

  if (config.importMap) {
    const importMapFile = join(dirname(file), config.importMap);
    console.log(importMapFile);

    await Deno.writeTextFile(
      importMapFile,
      JSON.stringify(importMap, null, 2) + "\n",
    );
    console.log(brightGreen("Import map file saved:"), importMapFile);
  }

  if (extname(file) === ".jsonc") {
    console.log(
      brightYellow("deno.jsonc needs to be manually updated:"),
      "Use deno.json to update it automatically",
      JSON.stringify(config, null, 2),
    );
  } else {
    await Deno.writeTextFile(file, JSON.stringify(config, null, 2) + "\n");
    console.log(brightGreen("Deno configuration file saved:"), file);
  }
}

export function isUrl(path: string): boolean {
  return !!path.match(/^(https?|file):\/\//);
}

export function createDate(str: string | number): Date | undefined {
  if (typeof str === "number") {
    return new Date(str);
  }

  const datetime = str.match(
    /^(\d{4})-(\d\d)-(\d\d)(?:-(\d\d)-(\d\d)(?:-(\d\d))?)?$/,
  );

  if (datetime) {
    const [, year, month, day, hour, minute, second] = datetime;

    return new Date(Date.UTC(
      parseInt(year),
      parseInt(month) - 1,
      parseInt(day),
      hour ? parseInt(hour) : 0,
      minute ? parseInt(minute) : 0,
      second ? parseInt(second) : 0,
    ));
  }

  if (str.match(/^\d+$/)) {
    return new Date(parseInt(str));
  }
}

export async function read(path: string, isBinary: true): Promise<Uint8Array>;
export async function read(path: string, isBinary: false): Promise<string>;
export async function read(
  path: string,
  isBinary: boolean,
): Promise<string | Uint8Array> {
  if (!isUrl(path)) {
    return isBinary ? Deno.readFile(path) : Deno.readTextFile(path);
  }

  const response = await fetch(path);

  return isBinary
    ? new Uint8Array(await response.arrayBuffer())
    : response.text();
}
