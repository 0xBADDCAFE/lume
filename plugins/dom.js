import { DOMParser } from "../deps/dom.js";
import { merge } from "../utils.js";

// Default options
const defaults = {
  extensions: [".html"],
  fn() {},
};

export default function (userOptions) {
  const parser = new DOMParser();
  const options = merge(defaults, userOptions);

  return (site) => {
    site.process(options.extensions, processor);

    function processor(page) {
      const document = parser.parseFromString(page.content, "text/html");
      options.fn(document, page);
      page.content = document.documentElement.outerHTML;
    }
  };
}
