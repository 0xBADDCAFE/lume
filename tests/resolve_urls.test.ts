import { assertStrictEquals as equals } from "../deps/assert.ts";
import { getSite, testPage } from "./utils.ts";
import resolveUrls from "../plugins/resolve_urls.ts";
import slugifyUrls from "../plugins/slugify_urls.ts";
import { Element } from "../deps/dom.ts";
import { Page } from "../core.ts";

Deno.test("relative_url plugin", async () => {
  const site = getSite({
    src: "resolve_urls",
  });

  site.copy("statics", "");
  site.use(resolveUrls());
  site.use(slugifyUrls()); // Test combined with slugify_urls
  await site.build();

  testPage(site, "/index", (page) => {
    equals(getHref(page, 0), "/bar");
    equals(getHref(page, 1), "foo");
    equals(getHref(page, 2), "./foo");
    equals(getHref(page, 3), "../foo");
    equals(getHref(page, 4), "#foo");
    equals(getHref(page, 5), "?foo=bar");
    equals(getHref(page, 6), "/other/");
    equals(getHref(page, 7), "/other/");
    equals(getHref(page, 8), "https://domain.com");
    equals(getHref(page, 9), "/other/?tab=1");
    equals(getHref(page, 10), "/other/#tab-1");
    equals(getHref(page, 11), "/");
    equals(getHref(page, 12), "/tilde-and-ene/");
  });

  testPage(site, "/articles/article-1", (page) => {
    equals(getHref(page, 0), "/drafts/second-article/");
    equals(getHref(page, 1), "/other/");
    equals(getHref(page, 2), "/");
  });

  testPage(site, "/articles/article-2", (page) => {
    equals(getHref(page, 0), "/articles/first-article/");
    equals(getHref(page, 1), "/other/");
    equals(getHref(page, 2), "/");
  });

  testPage(site, "/other", (page) => {
    equals(getHref(page, 0), "/robots.txt");
  });
});

function getHref(page: Page, pos: number) {
  const a = page.document?.querySelectorAll("a")[pos] as Element | null;

  if (!a) {
    return;
  }

  return a.getAttribute("href");
}
