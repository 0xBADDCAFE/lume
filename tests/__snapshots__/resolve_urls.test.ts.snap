export const snapshot = {};

snapshot[`relative_url plugin 1`] = `5`;

snapshot[`relative_url plugin 2`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
  ],
}
`;

snapshot[`relative_url plugin 3`] = `
[
  {
    entry: "/statics/robots.txt",
    flags: [],
    outputPath: "/robots.txt",
  },
]
`;

snapshot[`relative_url plugin 4`] = `
{
  content: "<!DOCTYPE html>
" +
    "<html><head></head><body><p>First article</p>
" +
    '<p><a href="/drafts/second-article/">Go to second</a>
' +
    '<a href="/other/">Go to other</a>
' +
    '<a href="/">Go to index</a></p>
' +
    "</body></html>",
  data: {
    children: true,
    content: "<!-- deno-fmt-ignore-file -->
" +
      "
" +
      "First article
" +
      "
" +
      "[Go to second](./article-2.md)
" +
      "[Go to other](../other.md)
" +
      "[Go to index](/index.md)",
    date: 1970-01-01T00:00:00.000Z,
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    url: "/articles/first-article/",
  },
  dest: {
    ext: ".html",
    path: "/articles/first-article/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/articles/article-1",
    remote: undefined,
    slug: "article-1",
  },
}
`;

snapshot[`relative_url plugin 5`] = `
{
  content: "<!DOCTYPE html>
" +
    "<html><head></head><body><p>Second article</p>
" +
    '<p><a href="/articles/first-article/">Go to first</a>
' +
    '<a href="/other/">Go to other</a>
' +
    '<a href="/">Go to index</a></p>
' +
    "</body></html>",
  data: {
    children: true,
    content: "<!-- deno-fmt-ignore-file -->
" +
      "
" +
      "Second article
" +
      "
" +
      "[Go to first](article-1.md)
" +
      "[Go to other](../other.md)
" +
      "[Go to index](/index.md)",
    date: 1970-01-01T00:00:00.000Z,
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    url: "/drafts/second-article/",
  },
  dest: {
    ext: ".html",
    path: "/drafts/second-article/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/articles/article-2",
    remote: undefined,
    slug: "article-2",
  },
}
`;

snapshot[`relative_url plugin 6`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html><head></head><body><p><a href="/bar">url</a>
' +
    '<a href="foo">url</a>
' +
    '<a href="./foo">url</a>
' +
    '<a href="../foo">url</a>
' +
    '<a href="#foo">url</a>
' +
    '<a href="?foo=bar">url</a>
' +
    '<a href="/other/">url</a>
' +
    '<a href="/other/">url</a>
' +
    '<a href="https://domain.com">url</a>
' +
    '<a href="/other/?tab=1">url</a>
' +
    '<a href="/other/#tab-1">url</a>
' +
    '<a href="/robots.txt">url</a>
' +
    '<a href="/tilde-and-ene/">url</a></p>
' +
    "</body></html>",
  data: {
    children: true,
    content: "<!-- deno-fmt-ignore-file -->
" +
      "[url](/bar)
" +
      "[url](foo)
" +
      "[url](./foo)
" +
      "[url](../foo)
" +
      "[url](#foo)
" +
      "[url](?foo=bar)
" +
      "[url](~/other.md)
" +
      "[url](other.md)
" +
      "[url](https://domain.com)
" +
      "[url](other.md?tab=1)
" +
      "[url](other.md#tab-1)
" +
      "[url](~/statics/robots.txt)
" +
      "[url](tílde-and-eñe.md)
",
    date: 1970-01-01T00:00:00.000Z,
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    url: "/",
  },
  dest: {
    ext: ".html",
    path: "/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/index",
    remote: undefined,
    slug: "index",
  },
}
`;

snapshot[`relative_url plugin 7`] = `
{
  content: "<!DOCTYPE html>
" +
    "<html><head></head><body><p>Other page</p>
" +
    '<p><a href="/robots.txt">See robots.txt</a></p>
' +
    "</body></html>",
  data: {
    children: true,
    content: "Other page

[See robots.txt](statics/robots.txt)
",
    date: 1970-01-01T00:00:00.000Z,
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    url: "/other/",
  },
  dest: {
    ext: ".html",
    path: "/other/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/other",
    remote: undefined,
    slug: "other",
  },
}
`;

snapshot[`relative_url plugin 8`] = `
{
  content: "<!DOCTYPE html>
" +
    "<html><head></head><body><p>Tilde and eñe</p>
" +
    "</body></html>",
  data: {
    children: true,
    content: "Tilde and eñe
",
    date: 1970-01-01T00:00:00.000Z,
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    url: "/tilde-and-ene/",
  },
  dest: {
    ext: ".html",
    path: "/tilde-and-ene/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/tílde-and-eñe",
    remote: undefined,
    slug: "tílde-and-eñe",
  },
}
`;
